### ISSUE TITLE
accessibility: let users specify alt text for images

### ISSUE THREAD

**Issue by xworld21** [brucemiller/LaTeXML#1452]
This came up in a [message from Peter Sweby](https://lists.informatik.uni-erlangen.de/pipermail/latexml/2020-February/002437.html): while `ltx:graphics` supports a `description` attribute which would be become `alt`, there is no way to access it from LaTeX. Alternative text is necessary to meet the Web Content Accessibility Guidelines [1]. Likewise, `<svg>` supports a `<description>` element.

Furthermore, images (both `<img>` and `<svg>`) support a `title` attribute. `ltx:graphics` does not have an equivalent, but that can be fixed easily.

The main trouble is, there is no LaTeX facility for adding alternative text, at best there is `\pdftooltip` which can be used to set the `title`.

I think a command in `latexml.sty` would be appropriate for this? I suppose it could be

- `\lxWithDesc{}{}` , `\lxWithTitle{}{}`, `\lxWithTitleDesc{}{}{}`: add title and description to the first `ltx:graphics` or SVG image produced by the second/third argument, throw a warning if there isn't one;
- `\lxAddDesc{}` and similar (although I should understand first how it interacts with `\begin{tikzpicture}` and the likes: should the command go inside the environment? can it go right after?).

The commands could be reused in other contexts, for instance if one wants to override the textual description of mathematical content (a matter for a subsequent issue, maybe).

PS: for reference, there is also `longdesc`, which is just a URI to a long description available somewhere else. WCAG still mentions it (see technique H45), but it is deprecated in HTML5 [2], the alternative being just sticking an `<a>` tag around the image.

[1] [Understanding Success Criterion 1.1.1: Non-text Content](https://www.w3.org/WAI/WCAG22/Understanding/non-text-content.html)

[2] [HTML Standard](https://html.spec.whatwg.org/multipage/obsolete.html#attr-img-longdesc)

**Comment by nxg** [comment]
Another option here is to add keywords to the `\includegraphics` command:

    \documentclass{article}
    \usepackage{graphicx}
    
    \makeatletter
    % Support a shortdescription={foo} keyword for \includegraphics.
    % There is also an `unknownkeysallowed` package option which can be given to `keyval`,
    % to tell it to ignore unknown keys silently.
    \def\theshortdescription{}
    \define@key{Gin}  % Gin is the `graphicx` label used for keyval by graphicx
               {shortdescription}
               {\typeout{short description: #1}\gdef\theshortdescription{#1}}
    \makeatother
    
    \begin{document}
    Here is an image:
    \includegraphics[angle=45,shortdescription={foo bar}]{dangerousbend.pdf}
    
    The description is [\theshortdescription].
    
    \end{document}

Note that this doesn't require any hacking of the `graphicx` package, but is fully external to it.

I see that `graphicx.sty.ltxml` does something along these lines, already, so it might be that the only puzzle would be working out how to arrange things so that the `shortdescription=FOO` keyword didn't cause a problem for users when it appears in a file processed by normal LaTeX.

I don't know (come to think of it), how LaTeXML normally advises users how to handle markup specific to LaTeXML (ie, markup which would cause an error if a file is processed by normal LaTeX).  But whatever mechanism there is, the above might be a way of installing no-op-but-don't-error support for extra graphics keywords.

Also, this is obviously specific to graphics/graphicx, and doesn't help with maths or, eg, tikz.

**Comment by xworld21** [comment]
> how LaTeXML normally advises users how to handle markup specific to LaTeXML (ie, markup which would cause an error if a file is processed by normal LaTeX)

People can use `\usepackage{latexml}` for that. In principle, latexml.sty could set a hook that adds `shortdescription`. But a generic mechanism is probably better here because of all the other ways to embed pictures, and indeed for maths as well.

**Comment by xworld21** [comment]
By the way, I have experimented with it a bit. I see are three sensible ways of attaching a `@description`:
- to the previous node (so `\includegraphics` followed by e.g. `\lxDescription{...}`); I have an implementation
- to the node containing the insertion point (so inside `\begin{tikzpicture}` and similar); this is identical to `\lxAddClass{}`
- to the first node of the second argument (i.e. `\lxWithDescription{description}{\includegraphics}`); this is identical to `\lxWithClass{}{}`

The missing ingredient, for me, is that `setAttribute` does not raise a warning if the model does not allow `@description`. So my current implementation of `\lxDescription{...}` might well fail silently if e.g. a user leaves an empty line.

**Comment by brucemiller** [comment]
This sounds like an interesting enhancement. We could define a set of generic accessibility attributes (such as title, description, ... long/short? others?) and add those to several/many/most elements, at least including `ltx:graphic`.  Then, as you suggest, add a minimal number of macros taking keyword arguments for the annotations in `latexml.sty.ltxml` to annotate (the current node, parent, previous,?). Where appropriate, like `\includegraphics`, the keywords could automatically be included (I would think that `\latexml.sty` could define them as noops for stock latex)

**Comment by dginev** [comment]
I'm tempted to also wonder out loud whether we should be drawing in some of the core LaTeX team members in discussing this, as there are ongoing accessibility efforts targeting the classic PDF workflow, as advertised in the latest TUG2021 conference. Especially for cases where we're extending macros with very long-lived traditions, such as `\includegraphics`, it would be nice to see if there can be some consensus strategy for both fronts, rather than having latexml.sty patch into graphic.sty.

Of course nothing is blocking us from developing our own custom macros to experiment with, which we should.

**Comment by nxg** [comment]
I think that's an excellent idea.

My impression is that the ‘accessibility’ work in the LaTeX core is so-far focused on the more specific (and hard) problem of getting PDF tags into the output in a robust way, rather than the broader problem of standardising alternate and descriptive text, so that LaTeXML might be able to lead this usefully.

As you doubtless know, there's currently quite a lot of LaTeX3 stuff moving into the LaTeX core, so it might be more important than it usually is to stick to standard interfaces, and resist the ever-present temptation to monkey-patch.
